pkg/document/json/tree.go

```diff
@@ -62,30 +77,47 @@ func NewTree(ctx *change.Context, tree *crdt.Tree) *Tree {
 	}
 }
 
-// Edit edits this tree with the given node.
-func (t *Tree) Edit(fromIdx, toIdx int, content *TreeNode) bool {
-	if fromIdx > toIdx {
-		panic("from should be less than or equal to to")
+// validateTextNode make sure that text node have non-empty string value
+func validateTextNode(treeNode TreeNode) error {
+	if len(treeNode.Value) == 0 {
+		return ErrEmptyTextNode
 	}
-	ticket := t.context.IssueTimeTicket()
 
-	var node *crdt.TreeNode
-	if content != nil {
-		var attributes *crdt.RHT
-		if content.Attributes != nil {
-			attributes = crdt.NewRHT()
-			for key, val := range content.Attributes {
-				attributes.Set(key, val, ticket)
+	return nil
+}
+
+// validateTextNode make sure that treeNodes consist of only one of the two types (text or element)
+func validateTreeNodes(treeNodes []*TreeNode) error {
+	firstTreeNodeType := treeNodes[0].Type
+
+	if firstTreeNodeType == index.DefaultTextType {
+		for _, treeNode := range treeNodes {
+			if treeNode.Type != index.DefaultTextType {
+				return ErrMixedNodeType
+			}
+
+			if err := validateTextNode(*treeNode); err != nil {
+				return err
 			}
 		}
-		node = crdt.NewTreeNode(crdt.NewTreePos(ticket, 0), content.Type, attributes, content.Value)
-		for _, child := range content.Children {
-			if err := buildDescendants(t.context, child, node); err != nil {
-				panic(err)
+
+	} else {
+		for _, treeNode := range treeNodes {
+			if treeNode.Type == index.DefaultTextType {
+				return ErrMixedNodeType
 			}
 		}
 	}
 
+	return nil
+}
+
+// Edit edits this tree with the given nodes.
+func (t *Tree) Edit(fromIdx, toIdx int, contents []*TreeNode) bool {
```

Comment:
How about using the variadic function to keep the interface simple?

https://www.digitalocean.com/community/tutorials/how-to-use-variadic-functions-in-go